internal/services/shared/schema.go

```diff
@@ -90,42 +90,35 @@ func SanityCheckExistingRelationships(
 				ctx,
 				qy,
 				qyErr,
-				"cannot delete Relation `%s` in Object Definition `%s`, as a Relationship references it", delta.RelationName, nsdef.Name)
+				"cannot delete relation `%s` in object definition `%s`, as a relationship references it", delta.RelationName, nsdef.Name)
 			if err != nil {
 				return err
 			}
 
-		case namespace.RelationDirectWildcardTypeRemoved:
-			qy, qyErr := rwt.ReverseQueryRelationships(
-				ctx,
-				datastore.SubjectsFilter{
-					SubjectType:        delta.WildcardType,
-					OptionalSubjectIds: []string{tuple.PublicWildcard},
-				},
-				options.WithResRelation(&options.ResourceRelation{
-					Namespace: nsdef.Name,
-					Relation:  delta.RelationName,
-				}),
-				options.WithReverseLimit(options.LimitOne),
-			)
-			err = ErrorIfTupleIteratorReturnsTuples(
-				ctx,
-				qy,
-				qyErr,
-				"cannot remove allowed wildcard type `%s:*` from Relation `%s` in Object Definition `%s`, as a Relationship exists with it",
-				delta.WildcardType, delta.RelationName, nsdef.Name)
-			if err != nil {
-				return err
+		case namespace.RelationAllowedTypeRemoved:
+			var optionalSubjectIds []string
+			var relationFilter datastore.SubjectRelationFilter
+			optionalCaveatName := ""
+
+			if delta.AllowedType.GetPublicWildcard() != nil {
+				optionalSubjectIds = []string{tuple.PublicWildcard}
+			} else {
+				relationFilter = datastore.SubjectRelationFilter{
+					NonEllipsisRelation: delta.RelationName,
+				}
+			}
+
+			if delta.AllowedType.GetRequiredCaveat() != nil {
+				optionalCaveatName = delta.AllowedType.GetRequiredCaveat().CaveatName
 			}
 
-		case namespace.RelationDirectTypeRemoved:
 			qy, qyErr := rwt.ReverseQueryRelationships(
 				ctx,
 				datastore.SubjectsFilter{
-					SubjectType: delta.DirectType.Namespace,
-					RelationFilter: datastore.SubjectRelationFilter{
-						NonEllipsisRelation: delta.RelationName,
-					},
+					SubjectType:        delta.AllowedType.Namespace,
+					OptionalSubjectIds: optionalSubjectIds,
+					RelationFilter:     relationFilter,
+					OptionalCaveatName: optionalCaveatName,
```

Comment:
If I understand this correctly, the goal here is to detect where either a `allowed type in a relation` was removed, or a `caveat in a relation` was removed. The error message does not seem to make distinction between a what is being removed (refers to type, regardless it's a definition or a caveat), but because the names are global, it should be unambiguous, but potentially confusing. Can we perhaps improve the error message so it says:

`cannot remove allowed [caveat / definition]` type

This information is present in the `Delta` but I'm not entirely sure it provides unambiguous information about whether it was the relation allowed type or allowed caveat.

Reply:
the error message will include the `with caveat` if it has that to make it clear it is the caveated one